proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,626 @@
+account changes.
+
+A new syscall will be added to update the PRAW fees:
+
+```
+fn set_program_rebatable_account_write_fee(
+  address: Pubkey, 
+  microlamports_per_requested_cu: u64
+) -> Result<()>;
+```
+
+The syscall will take as arguments the account on which fee should be updated
+and the microlamports per cu that should be charged and return a `Ok` if the
+changes were successful else will return `Err`.
+
+The accounts database will need to store the value of microlamports per
+requested CU for the account. Changes to update the PRAW fees could be done only
+by the account owner. All programs will need to implement special instructions
+so that the program admin can use this syscall to update PRAW fees on the
+required accounts.
+
+Setting the microlamports per CU to `0` disables PRAW fees. The maximum amount
+of microlamports per CU that could be set will be 2 ^ 42. Lets assume that the
+maximum CU required to update these fees on an account is 1,000 CU. So the
+maximum PRAW fees required to update an account (`M`) is
+
+```
+M = 2^42 * 1000 CU * 10^-6 lamports
+M = 4.398046511104 * 10^12 * 1000 * 10^-6
+M = 4.398046511104 SOLs
+```
+
+So this makes sure that admin does not lock itself out from changing the PRAW
+fees.
+
+### Syscall to read PRAW fees
+
+A new syscall will be added to read PRAW fees on an account.
+
+```
+fn get_program_rebatable_account_write_fee(
+  address: Pubkey
+) -> u64;
+```
+
+The syscall will take address as input and return microlamports per
+CU. It will return `0` if no PRAW is set on the address.
+
+### Syscall to Rebate
+
+The owner of the account can issue a rebate of the PRAW fees paid to write lock
+a specific account. The rebate issued will be transferred back to the payer at
+the end of the transaction if it is successfully executed. Similar to the set
+and read PRAW fees this will need to be exposed through a syscall so that
+programs can implement custom authorization and delegate the decision to
+composing programs.
+
+Simple rebate schemes will verify merely signers, e.g an oracle preventing 3rd
+parties from write-locking their price feed. Dexes will need more complex rebate
+schemes based on instruction sysvar introspection.
+
+Syscall definition of the rebate mechanism is:
+
+```
+fn rebate_program_rebatable_account_write_fee(
+  address: Pubkey, 
+  microlamports_per_requested_cu: u64
+);
+```
+
+Rebate takes account on which rebate is issued and the amount of microlamports
+per CU that needs to be rebated as input. It returns void but the transaction
+will fail for invalid inputs. In case of multiple rebates on the same account
+only the highest amount of rebate will be taken into account. The rebated amount
+is always the minimum of the largest call to
+`rebate_program_rebatable_account_write_fee()` and the PRAW fees on the
+account. If program rebates `U64::MAX` it means all the PRAW fees on the account
+are rebated. The rebate amount cannot be negative. It will return an error
```

Comment:
so, what happens N where U64::MAX > N > PRAW_fees_on_the_account is given? it's same as the PRAW fees on the account?

Reply:
Yes all the PRAW fees on account will be rebated.